kie-ci/src/test/java/org/kie/scanner/concurrent/ConcurrentBuildTest.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.kie.scanner.concurrent;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.drools.compiler.kie.builder.impl.InternalKieModule;
+import org.drools.core.util.FileManager;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.kie.api.KieServices;
+import org.kie.api.builder.ReleaseId;
+import org.kie.scanner.AbstractKieCiTest;
+import org.kie.scanner.KieMavenRepository;
+import org.kie.test.testcategory.TurtleTestCategory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.junit.Assert.assertTrue;
+import static org.kie.scanner.KieMavenRepository.getKieMavenRepository;
+
+@Category(TurtleTestCategory.class)
+public class ConcurrentBuildTest extends AbstractKieCiTest {
+    private static final Logger LOG = LoggerFactory.getLogger(ConcurrentBuildTest.class);
+
+    private FileManager fileManager;
+
+    @Before
+    public void setUp() throws Exception {
+        this.fileManager = new FileManager();
+        this.fileManager.setUp();
+        ReleaseId releaseId = KieServices.Factory.get().newReleaseId("org.kie", "scanner-test", "1.0-SNAPSHOT");
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        this.fileManager.tearDown();
+    }
+
+    // This is TurtleTest. You can run this test with -PrunTurtleTests
+    @Test
+    public void concurrentBuildWithDependency() throws Exception {
```

Comment:
I simulate the reported scenario and reproduce the deadlock with this test. However, this doesn't reproduce the deadlock in Drools 8 (upstream main), because ReleaseIdImpl usage is refactored and we no longer use `ReleaseIdImpl.setSnapshotVersion`, so the scenario is not reproducible. It's a good thing, but I will likely forward-port the same fix without this unit test.